04-onion-routing.md

```diff
@@ -755,59 +763,143 @@ Note that these error messages are not reliable, as they are not placed on-chain
 due to the possibility of hop failure.
 
 Intermediate hops store the shared secret from the forward path and reuse it to
-obfuscate any corresponding return packet during each hop.
+authenticate and obfuscate any corresponding return packet during each hop.
 In addition, each node locally stores data regarding its own sending peer in the
 route, so it knows where to return-forward any eventual return packets.
-The node generating the error message (_erring node_) builds a return packet
+
+The origin node signals to each node on the path that it supports the error
+format that is detailed below by setting the `resolution_format` tlv record in
+the forward packet to `1`. It must make sure that each node supports this format
+by observing the node feature `option_fat_error`.
+
+## Erring node
+
+The node generating the error message builds a return packet
 consisting of the following fields:
 
 1. data:
-   * [`32*byte`:`hmac`]
    * [`u16`:`failure_len`]
    * [`failure_len*byte`:`failuremsg`]
    * [`u16`:`pad_len`]
    * [`pad_len*byte`:`pad`]
+   * [`27*9*byte`:`payloads`]
+   * [`378*32*byte`:`hmacs`]
+
+Where `failuremsg` is defined below, and `pad` are the extra bytes used to
+conceal length.
+
+The field `payloads` contains a per-hop payload. The erring node puts its
+payload at the start of this array and zeroes out the rest. The size of the
+field is based on the maximum number of hops in a route (27) and the payload
+size (9 bytes).
+
+The per-hop payload consists of the following fields:
+1. data:
+   * [`byte`:`payload_type`]
+   * [`u64`:`hold_time_ms`]
```

Comment:
Maybe we should make this a fixed size block that contains tlv data. A zero-length tlv record with id 1 could be used to signal that this node is the error source.

The length of the fixed-size block could for example be 32 bytes. That would fit the hold time as a tlv record and leaves space for future extensions.

Unfortunately this space needs to be reserved for all 27 hops, so it adds close to 1 kb of extra data to the failure message.

Reply:
Possibly we can parameterize this and let the sender signal that block length, similar to https://github.com/lightning/bolts/pull/1044#discussion_r1069621260 for the number of hops.

Basically the sender decides the failure structure `max_hops x failure_data_length`

Reply:
Updated PR with parameterization.